### PR TITLE
Freeze and allow unanalyzable CommonJS symbols to be reexported across bundles

### PR DESCRIPTION
**Pull request by wbinnssmith**
This:

* [x] ~Implements a warning when using named imports on json assets (thanks @mischnic for the direction in symbol propagation)~
* [ ] Allows named, unanalyzable CommonJS (and therefore json) symbols to be reexported, addressing https://github.com/parcel-bundler/parcel/issues/5779

Test Plan:
* [x] ~Add integration test for warnings~
* [ ] Add integration test for reexporting

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 7.78s | -54.00ms |
| Cached | 484.00ms | +2.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 537.00ms | +71.00ms ⚠️ |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 293.00ms | -263.00ms 🚀 |
| dist/legacy/index.76ce4b0c.js | 1.12kb | +0.00b | 587.00ms | +48.00ms ⚠️ |
| dist/modern/index.b43fcc94.js | 1.12kb | +0.00b | 527.00ms | -408.00ms 🚀 |
| dist/modern/index.html | 701.00b | +0.00b | 521.00ms | -40.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 58.00ms | +3.00ms ⚠️ |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 85.00ms | +5.00ms ⚠️ |
| dist/modern/index.b43fcc94.js | 1.12kb | +0.00b | 92.00ms | +12.00ms ⚠️ |
| dist/modern/index.html | 701.00b | +0.00b | 91.00ms | +11.00ms ⚠️ |
| dist/modern/index.0bc656bc.css | 77.00b | +0.00b | 91.00ms | +11.00ms ⚠️ |

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 22.81s | -521.00ms |
| Cached | 761.00ms | +7.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 477.78kb | +0.00b | 7.52s | -383.00ms 🚀 |
| dist/NotFound.f3269ae7.js | 542.00b | +0.00b | 247.00ms | -32.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 477.78kb | +0.00b | 156.00ms | +29.00ms ⚠️ |
| dist/PermalinkedComment.0a19c7e4.js | 4.24kb | +0.00b | 156.00ms | +29.00ms ⚠️ |
| dist/UserProfile.b2960402.js | 1.78kb | +0.00b | 156.00ms | +29.00ms ⚠️ |
| dist/NotFound.f3269ae7.js | 542.00b | +0.00b | 113.00ms | +20.00ms ⚠️ |
| dist/logo.24c8bf9e.png | 274.00b | +0.00b | 126.00ms | +19.00ms ⚠️ |

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 2.65m | +17.22s ⚠️ |
| Cached | 3.16s | +39.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.59a13c6f.js | 2.46mb | -4.00b 🚀 | 1.27m | +16.01s ⚠️ |
| dist/pdfRenderer.f5a3f4a7.js | 1.11mb | -4.00b 🚀 | 1.27m | +37.46s ⚠️ |
| dist/editorView.ea13a2b7.js | 588.08kb | +0.00b | 29.88s | -22.02s 🚀 |
| dist/EmojiPickerComponent.c4358f76.js | 139.84kb | +0.00b | 9.22s | -779.00ms 🚀 |
| dist/card.df70c9b7.js | 54.67kb | +0.00b | 9.97s | +1.73s ⚠️ |
| dist/Modal.558fb5db.js | 39.59kb | +0.00b | 4.28s | +400.00ms ⚠️ |
| dist/component.ad27194c.js | 31.05kb | +0.00b | 2.03s | -1.51s 🚀 |
| dist/esm.e3419f0a.js | 27.83kb | +0.00b | 9.22s | +2.90s ⚠️ |
| dist/component.f25dcadb.js | 22.80kb | +0.00b | 4.27s | +408.00ms ⚠️ |
| dist/js.8167e2d5.js | 16.55kb | +0.00b | 4.27s | +408.00ms ⚠️ |
| dist/ui.15dd7acd.js | 14.30kb | +0.00b | 9.08s | +986.00ms ⚠️ |
| dist/workerHasher.bc6346e1.js | 11.90kb | +0.00b | 6.89s | +480.00ms ⚠️ |
| dist/card.3a4028bd.js | 5.74kb | +0.00b | 10.90s | +689.00ms ⚠️ |
| dist/EmojiPickerComponent.4991ddd4.js | 3.68kb | +0.00b | 8.82s | +2.42s ⚠️ |
| dist/png-chunks-extract.608f438d.js | 3.56kb | +0.00b | 4.27s | +408.00ms ⚠️ |
| dist/index.ece3a13e.css | 3.46kb | +0.00b | 16.32s | -22.14s 🚀 |
| dist/Modal.08282ab6.js | 3.23kb | +0.00b | 3.43s | +2.38s ⚠️ |
| dist/16.b6ca7fd0.js | 2.50kb | +0.00b | 2.03s | +975.00ms ⚠️ |
| dist/ResourcedEmojiComponent.5046c0ff.js | 2.19kb | +0.00b | 9.07s | +1.17s ⚠️ |
| dist/date.a69000aa.js | 1.97kb | +0.00b | 4.64s | +343.00ms ⚠️ |
| dist/feedback.facaac2e.js | 1.87kb | +0.00b | 8.91s | +3.33s ⚠️ |
| dist/16.a9ea9b41.js | 1.86kb | +0.00b | 4.40s | +250.00ms ⚠️ |
| dist/browser.8883c452.js | 1.85kb | +0.00b | 15.79s | +3.43s ⚠️ |
| dist/workerHasher.31326e89.js | 1.77kb | +0.00b | 6.86s | +470.00ms ⚠️ |
| dist/workerHasher.d607e0b8.js | 1.77kb | +0.00b | 12.49s | -3.18s 🚀 |
| dist/list-number.d4568b49.js | 1.69kb | +0.00b | 5.21s | +791.00ms ⚠️ |
| dist/status.6595d133.js | 1.69kb | +0.00b | 4.88s | +252.00ms ⚠️ |
| dist/code.efde0f48.js | 1.61kb | +0.00b | 4.54s | +250.00ms ⚠️ |
| dist/link.5be155be.js | 1.54kb | +0.00b | 5.21s | +793.00ms ⚠️ |
| dist/heading6.5a6082c8.js | 1.53kb | +0.00b | 5.00s | -582.00ms 🚀 |
| dist/heading3.1c92891c.js | 1.52kb | +0.00b | 4.98s | -3.97s 🚀 |
| dist/16.07ccad90.js | 1.51kb | +0.00b | 4.59s | +564.00ms ⚠️ |
| dist/16.02056ae6.js | 1.47kb | +0.00b | 4.59s | +564.00ms ⚠️ |
| dist/16.b7da7bde.js | 1.45kb | +0.00b | 2.03s | +975.00ms ⚠️ |
| dist/16.bdd61d2d.js | 1.41kb | +0.00b | 4.53s | +366.00ms ⚠️ |
| dist/heading5.88bc6249.js | 1.41kb | +0.00b | 4.98s | -597.00ms 🚀 |
| dist/expand.4a483eb0.js | 1.39kb | +0.00b | 8.91s | +3.32s ⚠️ |
| dist/16.fa5f411c.js | 1.36kb | +0.00b | 4.40s | +250.00ms ⚠️ |
| dist/16.2a21d3f4.js | 1.34kb | +0.00b | 2.03s | +975.00ms ⚠️ |
| dist/heading2.22de70fe.js | 1.34kb | +0.00b | 4.98s | +254.00ms ⚠️ |
| dist/16.d10db5a3.js | 1.32kb | +0.00b | 4.41s | +251.00ms ⚠️ |
| dist/mention.c44b48ed.js | 1.30kb | +0.00b | 4.78s | +301.00ms ⚠️ |
| dist/heading4.a2524c70.js | 1.29kb | +0.00b | 4.98s | -3.97s 🚀 |
| dist/Modal.44e980e0.js | 1.29kb | +0.00b | 4.27s | +408.00ms ⚠️ |
| dist/layout.81f2196c.js | 1.28kb | +0.00b | 5.21s | +793.00ms ⚠️ |
| dist/16.9e97e787.js | 1.28kb | +0.00b | 2.03s | +975.00ms ⚠️ |
| dist/16.acff901d.js | 1.27kb | +0.00b | 4.40s | +372.00ms ⚠️ |
| dist/16.828b958e.js | 1.26kb | +0.00b | 4.59s | +564.00ms ⚠️ |
| dist/16.b9b1643b.js | 1.26kb | +0.00b | 4.40s | +377.00ms ⚠️ |
| dist/divider.ed7ccd2e.js | 1.26kb | +0.00b | 4.64s | +298.00ms ⚠️ |
| dist/quote.b6d8bffb.js | 1.25kb | +0.00b | 4.88s | +252.00ms ⚠️ |
| dist/media-card-analytics-error-boundary.ff0177b3.js | 1.25kb | +0.00b | 10.90s | +2.78s ⚠️ |
| dist/action.01ec8983.js | 1.24kb | +0.00b | 4.53s | +234.00ms ⚠️ |
| dist/decision.c48f9f1e.js | 1.22kb | +0.00b | 4.64s | +343.00ms ⚠️ |
| dist/panel-warning.b5f3360d.js | 1.22kb | +0.00b | 4.88s | +374.00ms ⚠️ |
| dist/list.2dfe3f18.js | 1.19kb | +0.00b | 4.78s | +353.00ms ⚠️ |
| dist/heading1.cc82b077.js | 1.19kb | +0.00b | 6.40s | +1.75s ⚠️ |
| dist/16.90358f40.js | 1.18kb | +0.00b | 4.87s | +711.00ms ⚠️ |
| dist/panel-error.7454ebc0.js | 1.12kb | +0.00b | 4.78s | +269.00ms ⚠️ |
| dist/panel.fc70749c.js | 1.11kb | +0.00b | 4.88s | +282.00ms ⚠️ |
| dist/panel-success.d5752da0.js | 1.06kb | +0.00b | 4.78s | +275.00ms ⚠️ |
| dist/panel-note.6a71a1f2.js | 1.05kb | +0.00b | 4.78s | +269.00ms ⚠️ |
| dist/media-card-analytics-error-boundary.9a500f44.js | 1019.00b | +0.00b | 10.07s | +1.97s ⚠️ |
| dist/media-card-analytics-error-boundary.b884d580.js | 1019.00b | +0.00b | 12.48s | -3.16s 🚀 |
| dist/simpleHasher.79137cb4.js | 767.00b | +0.00b | 6.86s | +462.00ms ⚠️ |
| dist/simpleHasher.2c5206fd.js | 767.00b | +0.00b | 15.27s | +1.83s ⚠️ |
| dist/index.html | 119.00b | +0.00b | 2.51s | -157.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/editorView.ea13a2b7.js | 588.08kb | +0.00b | 609.00ms | -34.00ms 🚀 |
| dist/card.df70c9b7.js | 54.67kb | +0.00b | 594.00ms | -31.00ms 🚀 |
| dist/Modal.558fb5db.js | 39.59kb | +0.00b | 479.00ms | -81.00ms 🚀 |
| dist/component.ad27194c.js | 31.05kb | +0.00b | 463.00ms | -97.00ms 🚀 |
| dist/component.f25dcadb.js | 22.80kb | +0.00b | 485.00ms | -80.00ms 🚀 |
| dist/js.8167e2d5.js | 16.55kb | +0.00b | 479.00ms | -85.00ms 🚀 |
| dist/EmojiPickerComponent.4991ddd4.js | 3.68kb | +0.00b | 575.00ms | -31.00ms 🚀 |
| dist/png-chunks-extract.608f438d.js | 3.56kb | +0.00b | 485.00ms | -79.00ms 🚀 |
| dist/16.b6ca7fd0.js | 2.50kb | +0.00b | 464.00ms | -91.00ms 🚀 |
| dist/ResourcedEmojiComponent.5046c0ff.js | 2.19kb | +0.00b | 575.00ms | -31.00ms 🚀 |
| dist/date.a69000aa.js | 1.97kb | +0.00b | 521.00ms | -52.00ms 🚀 |
| dist/images.ece03ae5.js | 1.91kb | +0.00b | 532.00ms | -48.00ms 🚀 |
| dist/16.a9ea9b41.js | 1.86kb | +0.00b | 505.00ms | -64.00ms 🚀 |
| dist/16.89d45d15.js | 1.80kb | +0.00b | 492.00ms | -73.00ms 🚀 |
| dist/list-number.d4568b49.js | 1.69kb | +0.00b | 539.00ms | -41.00ms 🚀 |
| dist/status.6595d133.js | 1.69kb | +0.00b | 552.00ms | -36.00ms 🚀 |
| dist/code.efde0f48.js | 1.61kb | +0.00b | 517.00ms | -56.00ms 🚀 |
| dist/link.5be155be.js | 1.54kb | +0.00b | 536.00ms | -46.00ms 🚀 |
| dist/16.07ccad90.js | 1.51kb | +0.00b | 501.00ms | -67.00ms 🚀 |
| dist/16.00f25203.js | 1.47kb | +0.00b | 489.00ms | -76.00ms 🚀 |
| dist/16.02056ae6.js | 1.47kb | +0.00b | 494.00ms | -71.00ms 🚀 |
| dist/emoji.b5624cf5.js | 1.46kb | +0.00b | 532.00ms | -48.00ms 🚀 |
| dist/16.b7da7bde.js | 1.45kb | +0.00b | 464.00ms | -91.00ms 🚀 |
| dist/16.00100306.js | 1.44kb | +0.00b | 514.00ms | -59.00ms 🚀 |
| dist/16.bdd61d2d.js | 1.41kb | +0.00b | 514.00ms | -59.00ms 🚀 |
| dist/16.fa5f411c.js | 1.36kb | +0.00b | 509.00ms | -64.00ms 🚀 |
| dist/16.2a21d3f4.js | 1.34kb | +0.00b | 464.00ms | -91.00ms 🚀 |
| dist/16.d10db5a3.js | 1.32kb | +0.00b | 509.00ms | -64.00ms 🚀 |
| dist/16.c9db7e6b.js | 1.32kb | +0.00b | 517.00ms | -56.00ms 🚀 |
| dist/mention.c44b48ed.js | 1.30kb | +0.00b | 539.00ms | -41.00ms 🚀 |
| dist/Modal.44e980e0.js | 1.29kb | +0.00b | 479.00ms | -85.00ms 🚀 |
| dist/layout.81f2196c.js | 1.28kb | +0.00b | 536.00ms | -46.00ms 🚀 |
| dist/16.9e97e787.js | 1.28kb | +0.00b | 464.00ms | -91.00ms 🚀 |
| dist/16.acff901d.js | 1.27kb | +0.00b | 505.00ms | -64.00ms 🚀 |
| dist/16.015dfb9d.js | 1.26kb | +0.00b | 494.00ms | -71.00ms 🚀 |
| dist/16.828b958e.js | 1.26kb | +0.00b | 500.00ms | -68.00ms 🚀 |
| dist/16.b9b1643b.js | 1.26kb | +0.00b | 501.00ms | -68.00ms 🚀 |
| dist/16.ec84a250.js | 1.26kb | +0.00b | 489.00ms | -76.00ms 🚀 |
| dist/divider.ed7ccd2e.js | 1.26kb | +0.00b | 531.00ms | -49.00ms 🚀 |
| dist/quote.b6d8bffb.js | 1.25kb | +0.00b | 550.00ms | -38.00ms 🚀 |
| dist/action.01ec8983.js | 1.24kb | +0.00b | 518.00ms | -55.00ms 🚀 |
| dist/decision.c48f9f1e.js | 1.22kb | +0.00b | 522.00ms | -52.00ms 🚀 |
| dist/panel-warning.b5f3360d.js | 1.22kb | +0.00b | 546.00ms | -36.00ms 🚀 |
| dist/list.2dfe3f18.js | 1.19kb | +0.00b | 539.00ms | -41.00ms 🚀 |
| dist/heading1.cc82b077.js | 1.19kb | +0.00b | 560.00ms | -30.00ms 🚀 |
| dist/16.90358f40.js | 1.18kb | +0.00b | 509.00ms | -64.00ms 🚀 |
| dist/panel-error.7454ebc0.js | 1.12kb | +0.00b | 543.00ms | -38.00ms 🚀 |
| dist/panel.fc70749c.js | 1.11kb | +0.00b | 547.00ms | -43.00ms 🚀 |
| dist/table.77df779b.js | 1.10kb | +0.00b | 557.00ms | -32.00ms 🚀 |
| dist/panel-success.d5752da0.js | 1.06kb | +0.00b | 546.00ms | -36.00ms 🚀 |
| dist/panel-note.6a71a1f2.js | 1.05kb | +0.00b | 543.00ms | -38.00ms 🚀 |
| dist/simpleHasher.79137cb4.js | 767.00b | +0.00b | 575.00ms | -31.00ms 🚀 |

</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 19.52s | +305.00ms |
| Cached | 632.00ms | +2.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/293355270777602573)**

File: packages/core/core/src/requests/AssetGraphRequest.js [hunk 1]
```diff
@@ -319,6 +325,35 @@ export class AssetGraphBuilder {
               namespaceReexportedSymbols.add(exportSymbol);
             }
           }
+
+          // TODO: Move to symbol registration that happens in development too,
+          // likely added by https://github.com/parcel-bundler/parcel/pull/5528
+          if (
+            assetNode.value.meta.isJSON === true /* for Flow */ &&
```
Comment (devongovett) [inline packages/core/core/src/requests/AssetGraphRequest.js]: Should this really be in core? 😬

**Review by devongovett — COMMENTED** [review]
Why can't we actually support named imports from JSON? We do for CommonJS for example. Why is JSON different?

**Comment by wbinnssmith** [comment]
> Why can't we actually support named imports from JSON? We do for CommonJS for example. Why is JSON different?

For one, they're not a part of the JSON Modules spec: https://github.com/tc39/proposal-json-modules

**Comment by devongovett** [comment]
Heh well, we do a lot of stuff that's not standard (eg CJS). Feels like people will expect this to work?

**Comment by wbinnssmith** [comment]
> Feels like people will expect this to work?

Yeah, it feels intuitive. FWIW Webpack 5 is also warning on this and I imagine phasing it out in the long term.

**Comment by mischnic** [comment]
1. The default import is inconsistent with the named imports:
```js
import {default as a, b} from './test.json'; // {default: 1, b: 3}
console.log(a, b); //  `{default: 1, b: 3}, 3` instead of `1, 3`
```

2. This 
```js
export {a} from "./test.json";
```
is impossible to output as a pure ESM module, because `a` isn't actually a binding but it should still be live as normal named JSON imports would also be live?
```js
import {default as json, b} from './test.json'; // {a: 2, b: 3}
json.b = 2;
console.log(json, b); // `{a: 2, b: 2}, 2`
```
and doing 
```js
var _frozenA = json.a;
export {_frozenA as a};
```
might change the behaviour.

<br>

There's also a section here: https://github.com/tc39/proposal-json-modules#why-dont-json-modules-support-named-exports

**Comment by devongovett** [comment]
Hmm I was thinking the default import would correspond to the `default` property in the JSON file and you'd use `import *` to get the whole file. Seems like that would solve most of these issues, plus enable tree shaking. Plus all of this is just a nice shortcut, you could always use the namespace to get around any weirdness caused by ESM.

**Comment by mischnic** [comment]
> Hmm I was thinking the default import would correspond to the default property in the JSON file and you'd use import * to get the whole file.

That would not only be non-standard but even incompatible with the JSON modules standard.

**Comment by devongovett** [comment]
So is interop with CJS but we support that. Not sure why JSON is different - it's an object just like CJS. 🤷‍♂️

**Comment by mischnic** [comment]
It is equivalent to `modules.exports = {...json...}`, that's why `default` gives you the exports object itself and named imports return properties  (= interop).

The JSON module you're describing wouldn't behave like CJS.

**Comment by wbinnssmith** [comment]
This will be addressed in the upcoming scope hoisting implementation.

**Pull request closed without merging**